common/src/main/java/org/apache/celeborn/common/network/server/CreditStreamManager.java

```diff
@@ -0,0 +1,300 @@
+import org.slf4j.LoggerFactory;
+
+import org.apache.celeborn.common.meta.FileInfo;
+import org.apache.celeborn.common.network.server.memory.MemoryManager;
+import org.apache.celeborn.common.util.JavaUtils;
+
+public class CreditStreamManager {
+  private static final Logger logger = LoggerFactory.getLogger(CreditStreamManager.class);
+  private final AtomicLong nextStreamId;
+  private final ConcurrentHashMap<Long, StreamState> streams;
+  private final ConcurrentHashMap<FileInfo, MapDataPartition> activeMapPartitions;
+  private final HashMap<String, ExecutorService> storageFetcherPool = new HashMap<>();
+  private int definedMinReadBuffers;
+  private int definedMaxReadBuffers;
+  private int threadsPerMountPoint;
+  private int readAheadMin;
+  private final BlockingQueue<DelayedStreamId> recycleStreamIds = new DelayQueue<>();
+
+  @GuardedBy("lock")
+  private volatile Thread recycleThread;
+
+  private final Object lock = new Object();
+
+  public CreditStreamManager(
+      int minReadBuffers, int maxReadBuffers, int threadsPerMountpoint, int readAheadMin) {
+    nextStreamId = new AtomicLong((long) new Random().nextInt(Integer.MAX_VALUE) * 1000);
+    streams = JavaUtils.newConcurrentHashMap();
+    activeMapPartitions = JavaUtils.newConcurrentHashMap();
+    definedMinReadBuffers = minReadBuffers;
+    definedMaxReadBuffers = maxReadBuffers;
+    threadsPerMountPoint = threadsPerMountpoint;
+    this.readAheadMin = readAheadMin;
+    MemoryManager.instance().setCreditStreamManager(this);
+    logger.debug(
+        "Initialize buffer stream manager with {} {} {}",
+        definedMinReadBuffers,
+        definedMaxReadBuffers,
+        threadsPerMountpoint);
+  }
+
+  public long registerStream(
+      Consumer<Long> callback,
+      Channel channel,
+      int initialCredit,
+      int startSubIndex,
+      int endSubIndex,
+      FileInfo fileInfo)
+      throws IOException {
+    long streamId = nextStreamId.getAndIncrement();
+    StreamState streamState = new StreamState(channel, fileInfo.getBufferSize());
+    streams.put(streamId, streamState);
+    logger.debug(
+        "Register stream start from {}, streamId: {}, fileInfo: {}",
+        channel.remoteAddress(),
+        streamId,
+        fileInfo);
+    MapDataPartition mapDataPartition;
+    synchronized (activeMapPartitions) {
+      mapDataPartition = activeMapPartitions.get(fileInfo);
+      if (mapDataPartition == null) {
+        mapDataPartition =
+            new MapDataPartition(
+                definedMinReadBuffers,
+                definedMaxReadBuffers,
+                storageFetcherPool,
+                threadsPerMountPoint,
+                fileInfo,
+                id -> recycleStream(id),
+                readAheadMin);
+        activeMapPartitions.put(fileInfo, mapDataPartition);
+      }
+      mapDataPartition.addStream(streamId);
+    }
+
+    addCredit(initialCredit, streamId);
+    streamState.setMapDataPartition(mapDataPartition);
```

Comment:
This line should precedes addCredit since addCredit will getMapPartition from streamState. Just put this line before mapDataPartition.addStream(streamId);